coderd/prebuilds/preset_snapshot.go

```diff
@@ -83,6 +118,76 @@ func (ra *ReconciliationActions) IsNoop() bool {
 	return ra.Create == 0 && len(ra.DeleteIDs) == 0 && ra.BackoffUntil.IsZero()
 }
 
+// MatchesCron interprets a cron spec as a continuous time range,
+// and returns whether the provided time value falls within that range.
+func MatchesCron(cronExpression string, at time.Time) (bool, error) {
+	sched, err := cron.TimeRange(cronExpression)
+	if err != nil {
+		return false, xerrors.Errorf("failed to parse cron expression: %w", err)
+	}
+
+	return sched.IsWithinRange(at), nil
+}
+
+// CalculateDesiredInstances returns the number of desired instances based on the provided time.
+// If the time matches any defined prebuild schedule, the corresponding number of instances is returned.
+// Otherwise, it falls back to the default number of instances specified in the prebuild configuration.
+func (p PresetSnapshot) CalculateDesiredInstances(at time.Time) int32 {
+	if len(p.PrebuildSchedules) == 0 {
+		// If no schedules are defined, fall back to the default desired instance count
+		return p.Preset.DesiredInstances.Int32
+	}
+
+	// Validate that the provided timezone is valid
+	_, err := time.LoadLocation(p.Preset.SchedulingTimezone)
+	if err != nil {
+		p.logger.Error(context.Background(), "invalid timezone in prebuild scheduling configuration",
+			slog.F("preset_id", p.Preset.ID),
+			slog.F("timezone", p.Preset.SchedulingTimezone),
+			slog.Error(err))
+
+		// If timezone is invalid, fall back to the default desired instance count
+		return p.Preset.DesiredInstances.Int32
+	}
+
+	// Validate that all prebuild schedules are valid and don't overlap with each other.
+	// If any schedule is invalid or schedules overlap, fall back to the default desired instance count.
+	cronSpecs := make([]string, len(p.PrebuildSchedules))
+	for i, schedule := range p.PrebuildSchedules {
+		cronSpecs[i] = schedule.CronExpression
+	}
+	err = tf_provider_helpers.ValidateSchedules(cronSpecs)
+	if err != nil {
+		p.logger.Error(context.Background(), "schedules are invalid or overlap with each other",
+			slog.F("preset_id", p.Preset.ID),
+			slog.F("cron_specs", cronSpecs),
+			slog.Error(err))
+
+		// If schedules are invalid, fall back to the default desired instance count
+		return p.Preset.DesiredInstances.Int32
+	}
+
+	// Look for a schedule whose cron expression matches the provided time
+	for _, schedule := range p.PrebuildSchedules {
+		// Prefix the cron expression with timezone information
+		cronExprWithTimezone := fmt.Sprintf("CRON_TZ=%s %s", p.Preset.SchedulingTimezone, schedule.CronExpression)
+		matches, err := MatchesCron(cronExprWithTimezone, at)
+		if err != nil {
+			p.logger.Error(context.Background(), "cron expression is invalid",
+				slog.F("preset_id", p.Preset.ID),
+				slog.F("cron_expression", cronExprWithTimezone),
+				slog.Error(err))
+			continue
+		}
+		if matches {
+			return schedule.DesiredInstances
```

Comment:
I wonder if we should log that a schedule matched for debugging purposes.

Reply:
I can add this in a follow-up PR

Reply:
Why not just add it here? It's so simple to do.